hydromt/data_catalog.py

```diff
@@ -500,15 +614,28 @@ def to_dict(
                     ).replace("\\", "/")
             # remove non serializable entries to prevent errors
             source_dict = _process_dict(source_dict, logger=self.logger)  # TODO TEST
-            sources_out.update({name: source_dict})
+            if name in sources_out:
+                existing = sources_out.pop(name)
+                if existing == source_dict:
+                    sources_out.update({name: source_dict})
+                    continue
+                base, diff_existing, diff_new = partition_dictionaries(
+                    source_dict, existing
+                )
+                _ = base.pop("driver_kwargs", None)
+
+                base["variants"] = ([diff_new, diff_existing],)
+                sources_out[name] = base
+            else:
+                sources_out.update({name: source_dict})
         if meta:
             sources_out = {"meta": meta, **sources_out}
         return sources_out
 
     def to_dataframe(self, source_names: List = []) -> pd.DataFrame:
         """Return data catalog summary as DataFrame."""
         d = dict()
-        for name, source in self._sources.items():
+        for name, source in self.iter_sources():
             if len(source_names) > 0 and name not in source_names:
                 continue
             d[name] = source.summary()
```

Comment:
I'll see if I can do some testing, but in general pandas indexes don't have to be unique, so if it's just for printing, we can keep it as is.

Reply:
actually scratch that, I see what you mean now. I'll see if I can investigate the ramifications

Reply:
I went with the second variation, which seemed the most consistent to me, because you can still index like normal, and I think it is actually the intuitive behavior that if you slice into the data frame on the name of a source you should get back multiple rows, which should be the behavior under this impl.